### ISSUE TITLE
[FR] Plot metrics across multiple runs in an experiment

### ISSUE THREAD

**Issue by tanmaymathur89** [mlflow/mlflow#2768]
## Willingness to contribute
The MLflow Community encourages new feature contributions. Would you or another member of your organization be willing to contribute an implementation of this feature (either as an MLflow Plugin or an enhancement to the MLflow code base)?

- [ ] Yes. I can contribute this feature independently.
- [x] Yes. I would be willing to contribute this feature with guidance from the MLflow community.
- [ ] No. I cannot contribute this feature at this time.

## Proposal Summary
Allow Tracking server to plot metrics across runs with timestamp as the x-axis.

## Motivation
- What is the use case for this feature?
We use tracking server to log model artifacts, job metrics & metadata. Job metadata is logged in tags section. However, if we re-run the job and log information to a new run, tracking server does not allow the users to plot the metrics across multiple runs. This change will allow users to log metrics, and job metadata to a new run.

- Why is this use case valuable to support for MLflow users in general?
It extends the functionality of tracking server.

- Why is this use case valuable to support for your project(s) or organization?
We have automated workflows that run at a schedule. We want to save automated workflow information as a new run but also want the ability to see the metrics trends over time.

- Why is it currently difficult to achieve this use case? (please be as specific as possible about why related MLflow features and components are insufficient)
Tracking server cannot plot metrics across runs.

## Which MLflow component(s) does this feature affect?

- [x] UI
- [ ] Command Line Interface
- [ ] API
- [ ] Examples
- [ ] Docs
- [x] Tracking
- [ ] Projects
- [ ] Artifacts
- [ ] Models
- [ ] Model Registry
- [ ] Scoring
- [ ] R
- [ ] Java
- [ ] Python

**Comment by harupy** [comment]
@tanmaymathur89 

Is this what you're looking for?

![plot-multiple-metrics](https://user-images.githubusercontent.com/17039389/80664619-37135300-8ad2-11ea-85cc-03f3eeaf805c.gif)

**Comment by tanmaymathur89** [comment]
@harupy what version of mlflow is this?

**Comment by harupy** [comment]
@tanmaymathur89 

This is v1.8.0.

**Comment by tanmaymathur89** [comment]
@harupy Thank you! This should solve our issue.

**Comment by tanmaymathur89** [comment]
We have v1.7.2 but it only plots a bar chart with x-axis

**Comment by tanmaymathur89** [comment]
@harupy this made me realize that it actually plot two different curves in the same graph but it should be a one curve for a parameter.

**Comment by harupy** [comment]
Currently, it's not possible to plot metrics that belong to different runs as one curve.

**Comment by tanmaymathur89** [comment]
Can we add that feature?

**Comment by aarondav** [comment]
Would it be possible in your use-case to update the metrics for the existing run, using the [MlflowClient API](https://www.mlflow.org/docs/latest/python_api/mlflow.tracking.html)? Just wondering if this would avoid the need for special UI treatment to merge metrics from multiple runs together.

**Comment by tanmaymathur89** [comment]
@aarondav The reason we don't update the existing run is because it overwrites our workflow metadata too that are saved as tags in the run. 
Are there any alternatives to store workflow metadata and job metrics for multiple prediction runs?

**Comment by tanmaymathur89** [comment]
@aarondav any thoughts?

**Comment by smurching** [comment]
@tanmaymathur89 just to make sure we understand, it seems like your goal is the ability "to see the metrics trends over time" across runs, where the metric in question has a single relevant value per run (e.g. see how final model accuracy changes over time across different runs). One solution could be to log the run's start time as parameter or metric, and then rely on the scatter plot in the run comparison plot to view how your desired metric trends over time (i.e. plot the start time on the X axis and the relevant metric on the Y axis). It could make sense for us to enhance our existing run-comparison scatter plot to support plotting based on run start/end times on the X axis.

Note also that it's possible to create custom visualizations using the `mlflow.search_runs` API to load run data into a dataframe

**Comment by tanmaymathur89** [comment]
> @tanmaymathur89 just to make sure we understand, it seems like your goal is the ability "to see the metrics trends over time" across runs, where the metric in question has a single relevant value per run (e.g. see how final model accuracy changes over time across different runs). One solution could be to log the run's start time as parameter or metric, and then rely on the scatter plot in the run comparison plot to view how your desired metric trends over time (i.e. plot the start time on the X axis and the relevant metric on the Y axis). It could make sense for us to enhance our existing run-comparison scatter plot to support plotting based on run start/end times on the X axis.
> 
> Note also that it's possible to create custom visualizations using the `mlflow.search_runs` API to load run data into a dataframe

@smurching I dont believe you can log the start time as a [metric](https://www.mlflow.org/docs/latest/python_api/mlflow.html#mlflow.log_metric).Logging it as a parameter might work but we cannot enforce logging the start_time parameter.

It would be ideal if the graph could be plotted against the start time of the run since the information is already available.

Since we wan't it with in the UI, plotting it a pythonic way with mlflow api is possible however not easily accessible. We have a similar internal dashboard but want to switch to mlflow tracking server since all our current platform integrations work with tracking server. 

I can try creating a MR using mlflow start_time if thats the suggested solution.

**Comment by pepoandra** [comment]
@smurching @aarondav @tanmaymathur89 

Had this same issue & took a look at the code.

It seems that [the information about the runs](https://github.com/mlflow/mlflow/blob/master/mlflow/server/js/src/experiment-tracking/sdk/MlflowMessages.js#L113) (including `start_time`) is already available in the [CompareRunScatter.js component](https://github.com/mlflow/mlflow/blob/master/mlflow/server/js/src/experiment-tracking/components/CompareRunScatter.js#L17).

It seems to me that the work needed to add this feature is all in this one file, something along the lines of:

- Create a list of the plottable infos (Numerical values only) similar to[ these ones](https://github.com/mlflow/mlflow/blob/master/mlflow/server/js/src/experiment-tracking/components/CompareRunScatter.js#L33)
- Modify the [`GetValue` function](https://github.com/mlflow/mlflow/blob/master/mlflow/server/js/src/experiment-tracking/components/CompareRunScatter.js#L67) to accept this new type of value
- Add an option group for the Run Infos in the dropdown menus in [this line](https://github.com/mlflow/mlflow/blob/master/mlflow/server/js/src/experiment-tracking/components/CompareRunScatter.js#L190)

If you guys agree with this or can steer me in the right direction, I'm willing to work on a MR.

btw, I know this thread is 1 year old, but i'm sure more people are interested in this.

**Comment by pepoandra** [comment]
I could give it a shot.

**Comment by davidgilbertson** [comment]
Sorry if this isn't the place to ask, but I'm new to MLFlow, trying to work out how to compare metrics across multiple runs on a chart, _and have new runs added to the chart as they are created_. I've been through the docs but don't see much explaining the UI.

E.g. in TensorBoard, I can have this view, and now if I do another run with a different LR schedule, it will show up on the chart as it runs.
![image](https://user-images.githubusercontent.com/4443482/209734963-e4f7eb73-2cdd-40f9-b740-5d2082af38a9.png)


Is this possible with MLFlow? Or can I only see one metric at a time, and only compare runs that have been specifically selected before clicking the "Compare" button?

**Comment by benelot** [comment]
There are major improvements to the MLflow Experiments UI, including a configurable chart view providing visual model performance insights, a revamped parallel coordinates experience for tuning, and a streamlined table view with enhancements for search and filtering in the upcoming MLFlow version:
https://www.databricks.com/blog/2023/02/17/accelerate-your-model-development-new-mlflow-experiments-ui.html

**Comment by gachiemchiep** [comment]
@benelot glad to here that. it's amazing how MLflow is in 2.2 and still doesn't have an "add graph button".  I hope they can release the new MLflow version soon.
